homeassistant/components/control4/director_utils.py

```diff
@@ -15,21 +17,37 @@
 _LOGGER = logging.getLogger(__name__)
 
 
-async def director_update_data(
-    hass: HomeAssistant, entry: ConfigEntry, var: str
-) -> dict:
-    """Retrieve data from the Control4 director for update_coordinator."""
-    # possibly implement usage of director_token_expiration to start
-    # token refresh without waiting for error to occur
+async def _update_variables_for_config_entry(
+    hass: HomeAssistant, entry: ConfigEntry, variable_names: Set[str]
+) -> dict[int, dict[str, bool | int | str | dict]]:
+    """Retrieve data from the Control4 director."""
+    director = hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR]
+    data = await director.getAllItemVariableValue(variable_names)
+    result_dict: defaultdict[int, dict[str, bool | int | str | dict]] = defaultdict(
+        dict
+    )
+    for item in data:
+        typ = item.get("type", None)
+        value = item["value"]
+        if typ == "Boolean":
+            value = bool(int(value))
+        elif typ == "Number":
+            value = float(value)
+
+        result_dict[int(item["id"])][item["varName"]] = value
+    return dict(result_dict)
```

Comment:
I find all these force conversions confusing. What does the input look like?

Reply:
The input is a json then converted to a python dictionary. Control4 is a bit odd and will sometimes send these variables with various fields. These indicate the type of the variable and the value which is often a string. When a type is not specified we assume that the variable is type string or whatever type is supplied in json such as a number.

Reply:
Here is an example:

[{'id': 36, 'varName': 'LIGHT_LEVEL', 'type': 'Number', 'hidden': 0, 'value': 2, 'variableId': 1014, 'name': 'Card Access Wireless Motion Sensor', 'roomName': 'Theater Room'}, {'id': 39, 'varName': 'LIGHT_STATE', 'type': 'Boolean', 'hidden': 0, 'value': 0, 'variableId': 1000, 'name': 'Theater Lights', 'roomName': 'Theater Room'}, {'id': 39, 'varName': 'LIGHT_LEVEL', 'type': 'Number', 'hidden': 0, 'value': 0, 'variableId': 1001, 'name': 'Theater Lights', 'roomName': 'Theater Room'}, {'id': 984, 'varName': 'LIGHT_STATE', 'type': 'Boolean', 'hidden': 0, 'value': 0, 'variableId': 1000, 'name': 'Garage Lights', 'roomName': 'Garage'}, {'id': 991, 'varName': 'LIGHT_STATE', 'type': 'Boolean', 'hidden': 0, 'value': 0, 'variableId': 1000, 'name': 'Wine Light', 'roomName': 'Kitchen'}, {'id': 992, 'varName': 'LIGHT_STATE', 'type': 'Boolean', 'hidden': 0, 'value': 0, 'variableId': 1000, 'name': 'Christmas Tree', 'roomName': 'Kitchen'}, {'id': 996, 'varName': 'LIGHT_STATE', 'type': 'Boolean', 'hidden': 0, 'value': 0, 'variableId': 1000, 'name': 'Hallway Light', 'roomName': 'Great Room'}, {'id': 1000, 'varName': 'LIGHT_STATE', 'type': 'Boolean', 'hidden': 0, 'value': 0, 'variableId': 1000, 'name': 'Anay Bedroom Light', 'roomName': 'Anay Bedroom'}, {'id': 1002, 'varName': 'LIGHT_STATE', 'type': 'Boolean', 'hidden': 0, 'value': 0, 'variableId': 1000, 'name': 'Stairway Light', 'roomName': 'Game Room'}, {'id': 1004, 'varName': 'LIGHT_STATE', 'type': 'Boolean', 'hidden': 0, 'value': 0, 'variableId': 1000, 'name': 'Nalin Bedroom Light', 'roomName': 'Nalin Bedroom'}, {'id': 1027, 'varName': 'LIGHT_STATE', 'type': 'Boolean', 'hidden': 0, 'value': 0, 'variableId': 1000, 'name': 'String Lights', 'roomName': 'Covered Patio'}, {'id': 1032, 'varName': 'LIGHT_STATE', 'type': 'Boolean', 'hidden': 0, 'value': 0, 'variableId': 1000, 'name': 'Office Light', 'roomName': 'Study'}, {'id': 1034, 'varName': 'LIGHT_STATE', 'type': 'Boolean', 'hidden': 0, 'value': 0, 'variableId': 1000, 'name': 'Game Room Light', 'roomName': 'Game Room'}]